### PR TITLE
Fix URL encoding issue for filenames with special characters

### PR DESCRIPTION
**Pull request by jacksgt**
Hey,

as discussed in #7 , I have ported over the path encoding function from minio-go.
Before that, I also added some tests to check the new & correct behavior.
I have run these tests against AWS S3 buckets (and they were successful), I still want to try it out with a Minio instance (but I didn't get around to it yet).
Once I'm done with that, I will squash the commits.

Feedback welcome. :-)

### CODE REVIEW

File: simples3.go [hunk 1]
```diff
@@ -160,16 +163,20 @@ func (s3 *S3) getClient() *http.Client {
 }
 
 func (s3 *S3) getURL(bucket string, args ...string) (uri string) {
+	path := bucket
```
Comment (rhnvrm) [inline simples3.go:166]: Since this is an internal function, we can rename inside the function param itself to path.

File: simples3.go [hunk 2]
```diff
@@ -391,3 +398,45 @@ func (s3 *S3) FileDelete(u DeleteInput) error {
 
 	return nil
 }
+
+// if object matches reserved string, no need to encode them
+var reservedObjectNames = regexp.MustCompile("^[a-zA-Z0-9-_.~/]+$")
+
+// EncodePath encode the strings from UTF-8 byte representations to HTML hex escape sequences
```
Comment (rhnvrm) [inline simples3.go:405]: /s/EncodePath/encodePath

File: simples3.go [hunk 3]
```diff
@@ -160,16 +163,20 @@ func (s3 *S3) getClient() *http.Client {
...
+	path := bucket
+	if len(args) > 0 {
+		path += "/" + strings.Join(args, "/")
+	}
+	// need to encode special characters in the path part of the URL
+	encodedPath := encodePath(path)
+
 	if len(s3.Endpoint) > 0 {
-		uri = s3.Endpoint + "/" + bucket
+		uri = s3.Endpoint + encodedPath
```
Comment (rhnvrm) [inline simples3.go:174]: The TestCustomEndpoint unit test seems to be failing, could you please check it?

```
--- FAIL: TestCustomEndpoint (0.00s)
    simples3_test.go:340: S3.SetEndpoint() got = https://example.com
    simples3_test.go:346: S3.SetEndpoint() got = http://localhost:9000
    simples3_test.go:352: S3.SetEndpoint() got = https://example.com
```

**Review by rhnvrm — CHANGES_REQUESTED** [review]
Hey, a few suggestions...

**Comment by jacksgt** [comment]
Hey, thank you for the review!
I hope I addresses all the issues. :-)

Still need to test with my Minio instance, though...

**Comment by rhnvrm** [comment]
Hey

I just tested this with a local minio setup using their release binary and tests seem to pass. 

Documenting steps for future, I think we can use this for setting up automated testing. 

Had to create a testbucket from the UI (most likely this can be done using the cli). After that used the following env vars and ran `go test`.

```
AWS_S3_BUCKET=testbucket 
AWS_S3_ACCESS_KEY=minioadmin
AWS_S3_SECRET_KEY=minioadmin
AWS_S3_ENDPOINT=http://127.0.0.1:9000
```

**Comment by jacksgt** [comment]
Adding integration tests in the CI with a local Minio server is a great idea! Let's tackle that in a separate PR.

I have squashed the commits, so from my side this is ready to be merged.

**Review by rhnvrm — APPROVED** [review]
LGTM :+1:

**Comment by jacksgt** [comment]
Great, thanks for taking the time!

**Pull request merged**